### PR TITLE
fix: prometheusIngester headers from env

### PR DESCRIPTION
**Pull request by FuFikCZ**
original implementation counted with viper using yaml library and UnmarshalYAML interface. It was naive, viper uses
mapstructure instead of yaml package

### CODE REVIEW

File: pkg/prometheus_ingester/prometheus_ingester_test.go [hunk 1]
```diff
@@ -721,133 +720,134 @@ func Test_processHistogramIncrease(t *testing.T) {
...
+			name: "empty headers",
+			hs:   httpHeaders{},
+			want: map[string]string{},
 		},
 		{
-			name:    "none of valueFromString or valueFromEnv is set",
-			data:    []byte("name: name"),
-			wantErr: true,
+			name: "multiple headers",
+			hs:   httpHeaders{{Name: "header1", Value: &headerValue}, {Name: "header2", Value: &headerValue2}},
```
Comment (rudo-thomas) [inline pkg/prometheus_ingester/prometheus_ingester_test.go:740]: What I'm missing in these test is the whole path from YAML through viper into the structure. (Previously, the tests had the same problem and thus didn't catch the bug you're fixing in this PR.)

I suggest the tests have YAML in them, then something like the following is used to fill HttpHeaders:
```go
	var s struct { HH httpHeaders } // the structure we are filling
	in := `hh: [...]`
	v := viper.New()
	v.SetConfigType("yaml")
	if err := v.ReadConfig(strings.NewReader(in)); err != nil {
		panic(err)
	}
	if err := v.UnmarshalExact(&s); err != nil {
		panic(err)
	}
```

**Review by rudo-thomas — CHANGES_REQUESTED** [review]
The test need some love :)

File: pkg/prometheus_ingester/prometheus_ingester_test.go (hunk 1 above)
Comment (FuFikCZ) [reply pkg/prometheus_ingester/prometheus_ingester_test.go:740]: previous test had problem because of usage of custom unmarshaller. Unmarshaling yaml into struct using viper is already done by viper tests https://github.com/spf13/viper/blob/master/viper_test.go#L506

File: pkg/prometheus_ingester/prometheus_ingester.go [hunk 2]
```diff
@@ -106,31 +96,41 @@ func (h *httpHeader) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
-	return nil
-
+	return *h.Value, nil
 }
 
 type httpHeaders []httpHeader
 
-func (hs httpHeaders) toMap() map[string]string {
+func (hs httpHeaders) toMap() (map[string]string, error) {
 	httpHeaders := map[string]string{}
```
[the remark this answers is not in the document]
Comment (FuFikCZ) [reply pkg/prometheus_ingester/prometheus_ingester.go]: fixed in e046c468763412d3eba78d5e5a1a7be0be4e14c3

File: pkg/prometheus_ingester/prometheus_ingester.go [hunk 3]
```diff
@@ -77,27 +77,17 @@ type queryOptions struct {
...
-	type httpHeaderCustom httpHeader
-	var out httpHeaderCustom
-
-	if err := unmarshal(&out); err != nil {
-		return err
-	}
-
-	*h = httpHeader(out)
-
+func (h *httpHeader) validate() error {
```
[the remark this answers is not in the document]
Comment (FuFikCZ) [reply pkg/prometheus_ingester/prometheus_ingester.go]: fixed in e046c468763412d3eba78d5e5a1a7be0be4e14c3

File: pkg/prometheus_ingester/prometheus_ingester_test.go (hunk 1 above)
Comment (rudo-thomas) [reply pkg/prometheus_ingester/prometheus_ingester_test.go:740]: That's what I'm saying: The tests did not verify that the values got from the YAML all the way to the structure using viper. (Think unit test.)

These tests have the same problem: They currently test that once the data is in httpHeaders, it gets to the map[string]string output as expected. (Think unit test.) There is no test that verifies the YAML to httpHeaders path -- httpHeader values in the tests are all literals.

In other words, the current tests are prone to essentially the same problem we hit previously.

What I'm suggesting is to test that YAML input, when processed by viper, produces the map[string]string output as expected. (Think integration test.)

It should be a mechanical change, with the little helper function I have provided.

Comment (FuFikCZ) [reply pkg/prometheus_ingester/prometheus_ingester_test.go]: fixed b25af1336dbebc524011d8161541652b759b1788

**Review by rudo-thomas — APPROVED** [review]
